### PR TITLE
IGAPP-951: Add filtering to shelter offers

### PR DESCRIPTION
**Pull request by f1sh1918**
This pull request belongs to an issue on our [bugtracker](https://issues.integreat-app.de/).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword **IGAPP**.

Added minBeds & pets filter
http://localhost:9000/augsburg/de/offers/wohnraum

### CODE REVIEW

File: api-client/src/endpoints/createShelterEndpoint.ts [hunk 1]
```diff
@@ -6,14 +6,38 @@ import { JsonShelterType } from '../types'
 const SHELTER_URL = 'wohnraum.tuerantuer.org/wp-json/accommodations'
 
 export const SHELTER_ENDPOINT_NAME = 'shelter'
-type Params = { type: 'detail'; id: string; cityCode: string } | { type: 'list'; page: number; cityCode: string }
+type Params =
+  | { type: 'detail'; id: string; cityCode: string }
+  | { type: 'list'; page: number; cityCode: string; filter?: FilterProps }
```
Comment (klinzo) [inline api-client/src/endpoints/createShelterEndpoint.ts:11]: Does this have to be optional? I think logic would be a little bit easier if it is not and it should always be possible to set filters doing nothing as default. I think it is also never null/undefined.

File: api-client/src/endpoints/createShelterEndpoint.ts [hunk 2]
```diff
@@ -6,14 +6,38 @@ import { JsonShelterType } from '../types'
...
+    .join('')
 
 export default (): Endpoint<Params, ShelterModel[]> =>
   new EndpointBuilder<Params, ShelterModel[]>(SHELTER_ENDPOINT_NAME)
     .withParamsToUrlMapper(params => {
       const baseUrl = `https://${params.cityCode}.${SHELTER_URL}`
       if (params.type === 'list') {
-        return `${baseUrl}/list?page=${params.page}`
+        const filterParams = params.filter ? `${getFilterQueryParams(params.filter)}` : ''
+        return `${baseUrl}/list?page=${params.page}${filterParams}`
```
Comment (klinzo) [inline api-client/src/endpoints/createShelterEndpoint.ts:40]: Optional oc but seems a lot simpler to me.
```suggestion
        const { beds, pets } = params.filter
        let filterParams = beds ? `&min_beds=${beds}` : ''
        if (pets) {
          filterParams = `${filterParams}&pets=${pets}`
        }
        return `${baseUrl}/list?page=${params.page}${filterParams}`
```

File: translations/translations.json [hunk 3]
```diff
@@ -6126,7 +6126,9 @@
       "shelterType": "Art des Wohnraums",
       "startDate": "frühestens verfügbar",
       "availableBeds": "verfügbare Betten",
-      "duration": "mögliche Aufenthaltsdauer"
+      "duration": "mögliche Aufenthaltsdauer",
+      "facetBeds": "min. Anzahl Betten",
```
Comment (klinzo) [inline translations/translations.json:6130]: What is a facet? :see_no_evil: I think `Anzahl Betten` is enough, you also do that for english already.

File: translations/translations.json [hunk 4]
```diff
@@ -6126,7 +6126,9 @@
...
+      "facetPets": "Haustiere erlaubt"
```
Comment (klinzo) [inline translations/translations.json:6131]: We already have that translation. Doesn't hurt to have it twice though

File: web/src/components/FacetInput.tsx [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  flex-wrap: wrap;
+`
+
+const StyledInput = styled.input`
+  font-size: 14px;
+  margin-left: -1px;
+  border-radius: 0.25rem;
+  border-top-left-radius: 0;
+  border-bottom-left-radius: 0;
+  background-color: #fff;
```
Comment (klinzo) [inline web/src/components/FacetInput.tsx:15]: Maybe we could use colors from theme as much as possible.

File: web/src/components/FacetInput.tsx [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+const FacetInput: React.FC<FacetInputProps> = ({
+  value,
+  icon,
+  altTag,
+  updateSearchFilter,
+  name,
+  placeholder
+}: FacetInputProps): ReactElement => (
+  <FilterFacet>
+    <FilterIcon>
```
Comment (klinzo) [inline web/src/components/FacetInput.tsx:50]: No tooltip here?

File: web/src/components/InfiniteScrollList.tsx [hunk 7]
```diff
@@ -62,14 +63,14 @@ const InfiniteScrollList = <T,>({
...
-    if (resetOnLanguageChange) {
+    if (resetOnLanguageChange || filterValues) {
       setData([])
       setError(null)
       setLoading(false)
       setHasMore(true)
       setPage(defaultPage)
     }
-  }, [languageCode, resetOnLanguageChange, defaultPage])
+  }, [languageCode, resetOnLanguageChange, defaultPage, filterValues])
```
Comment (klinzo) [inline web/src/components/InfiniteScrollList.tsx:73]: Does this work? I think we could just use the dependency on `[loadData]`, that function should change on language change/filter value change already.

**Review by klinzo — COMMENTED** [review]
Everything is optional.

File: translations/translations.json (hunk 3 above)
Comment (f1sh1918) [reply translations/translations.json:6130]: facet is the regular name for a filter

File: web/src/components/FacetInput.tsx (hunk 6 above)
Comment (f1sh1918) [reply web/src/components/FacetInput.tsx:50]: i think the placeholder says enough

File: web/src/components/InfiniteScrollList.tsx (hunk 7 above)
Comment (f1sh1918) [reply web/src/components/InfiniteScrollList.tsx:73]: you have to reset the data first and then the load gets triggered auto (on filter changed)

File: web/src/routes/ShelterPage.tsx [hunk 8]
```diff
@@ -20,10 +22,17 @@ const ShelterPage = ({ cityModel, cityCode, languageCode, pathname, languages }:
   const { shelterId } = useParams()
   const { viewportSmall } = useWindowDimensions()
   const { t } = useTranslation('shelter')
+  const [filter, setFilter] = useState<FilterProps>({ beds: null, pets: null })
```
[the remark this answers is not in the document]
Comment (f1sh1918) [reply web/src/routes/ShelterPage.tsx:25]: i think if we get more filter, an filter object is the better way than to set for each one a new variable

File: api-client/src/endpoints/createShelterEndpoint.ts (hunk 2 above)
Comment (f1sh1918) [reply api-client/src/endpoints/createShelterEndpoint.ts:40]: if we get more filters, you want to add every filter here? Don't think thats an improvement to be honest.

Comment (klinzo) [reply api-client/src/endpoints/createShelterEndpoint.ts:40]: yep, would do it that way, we will never have more then five filters. but your way is fine so just keep it.

File: web/src/routes/ShelterPage.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (klinzo) [reply web/src/routes/ShelterPage.tsx:25]: yep, would do it that way, we will never have more then five filters. but your way is fine so just keep it.

Comment (klinzo) [inline web/src/routes/ShelterPage.tsx:25]: This default should be fine, then you don't need to check whether params are defined in the endpoint. At least it makes no sense to search for less than one bed.
```suggestion
  const [filter, setFilter] = useState<FilterProps>({ beds: '1', pets: '0' })
```

File: api-client/src/endpoints/createShelterEndpoint.ts (hunk 1 above)
Comment (f1sh1918) [reply api-client/src/endpoints/createShelterEndpoint.ts:11]: thats true, check last commit

File: web/src/components/InfiniteScrollList.tsx (hunk 7 above)
Comment (klinzo) [inline web/src/components/InfiniteScrollList.tsx:73]: This works and is way more future proof/easier to understand:
``` typescript
  useEffect(() => () => {
    setData([])
    setError(null)
    setLoading(false)
    setHasMore(true)
    setPage(defaultPage)
  }, [loadPage, defaultPage])
```

**Review by klinzo — APPROVED** [review]
Tested on firefox. Nice job :)

This still looks kind of active, maybe you can ask nikola for some better color scheme/highligting?
image

File: web/src/components/InfiniteScrollList.tsx (hunk 7 above)
Comment (klinzo) [reply web/src/components/InfiniteScrollList.tsx:73]: see suggestion below

File: translations/translations.json (hunk 3 above)
Comment (klinzo) [reply translations/translations.json:6141]: Never heard that name, but fine. I'd just have called it filter.

File: web/src/components/InfiniteScrollList.tsx (hunk 7 above)
Comment (f1sh1918) [reply web/src/components/InfiniteScrollList.tsx]: fixed

File: web/src/routes/ShelterPage.tsx (hunk 8 above)
Comment (f1sh1918) [reply web/src/routes/ShelterPage.tsx:25]: i would like to keep the placeholder in the input field --> rejected

File: translations/translations.json (hunk 3 above)
Comment (f1sh1918) [reply translations/translations.json:6141]: you can also rename it when we have checked ukrainian translations if you prefer 'filter'. i'm fine with that

Comment (klinzo) [reply translations/translations.json:6141]: no its fine, just had never heard that. but has no benefit so lets just keep it

**Review by Taggotty — APPROVED** [review]
Tested with chrome. Would be nice if you could add a test for filtering though. Especially as currently there aren't any shelters created in the cms. Or am I missing something 🙈

**Comment by f1sh1918** [comment]
> Tested with chrome. Would be nice if you could add a test for filtering though. Especially as currently there aren't any shelters created in the cms. Or am I missing something 🙈

we gonna create the tiles later. use link above. we won't add test, since this is only temp implementation

**Comment by klinzo** [comment]
> > Tested with chrome. Would be nice if you could add a test for filtering though. Especially as currently there aren't any shelters created in the cms. Or am I missing something see_no_evil
> 
> we gonna create the tiles later. use link above. we won't add test, since this is only temp implementation

Wohnraum offer just not activated in the test cms yet. Should be available shortly though after Sven has activated it. Link above should still work.

**Comment by klinzo** [comment]
> Tested with chrome. Would be nice if you could add a test for filtering though. Especially as currently there aren't any shelters created in the cms. Or am I missing something see_no_evil

@Taggotty offer tile should now also be available in test cms.

**Pull request merged**